### PR TITLE
policy: Remove denied identities maps

### PR DESCRIPTION
**Pull request by ianvernon**
Now that requirement selectors are folded into the selectors of the
rules in selector policies, we no longer need to keep track of
identities denied at L3 to produce the correct L4 policies. Removal of
DeniedIngressIdentities and DeniedEgressIdentities from the
EndpointPolicy also allows the removal of the matchingRules members
from the SelectorPolicy.

The generation of a SelectorPolicy is made faster by keeping note of
the selected rules on the first pass, and then iterating only over
those rules in subsequent passes. Note that this changes the set of
rules the later passes process and may result in different policy
trace output than before.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off by: Ian Vernon <ian@cilium.io>

This is cherry-picked from https://github.com/cilium/cilium/pull/7958 to reduce the size of said PR.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7962)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

File: pkg/policy/api/egress.go [hunk 1]
```diff
@@ -176,6 +180,29 @@ func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
 	return e.aggregatedSelectors
 }
 
+// GetDestinationEndpointSelectorsWithRequirements returns a slice of endpoints selectors covering
+// all L3 source selectors of the ingress rule
```
Comment (joestringer) [inline pkg/policy/api/egress.go:184]: I suspect that with the rest of the other PR eliminating dead code, we might be able to fold `GetDestinationEndpointSelectorsWithRequirements()` and `GetDestinationEndpointSelectors()` together?

Also:
```suggestion
// all L3 destination selectors of the egress rule
```

File: pkg/policy/repository_test.go [hunk 2]
```diff
@@ -1844,7 +1844,7 @@ Ingress verdict: allowed
 	ctx = buildSearchCtx("bar", "foo", 0)
 	expectedOut = `
 Resolving ingress policy for [any:foo]
-0/1 rules selected
+0/0 rules selected
```
Comment (joestringer) [inline pkg/policy/repository_test.go]: There aren't 0 rules in the repository, though?

I think we just need to count the total early on, and print that here. the `0/1` is correct (no rules select 'foo').

File: pkg/policy/repository.go [hunk 3]
```diff
@@ -143,12 +143,6 @@ func wildcardL3L4Rule(proto api.L4Proto, port int, endpoints api.EndpointSelecto
 	}
 }
 
-// wildcardL3L4Rules updates each ingress L7 rule to allow at L7 all traffic that
-// is allowed at L3-only or L3/L4.
-func (p *Repository) wildcardL3L4Rules(ctx *SearchContext, ingress bool, l4Policy L4PolicyMap) {
```
Comment (joestringer) [inline pkg/policy/repository.go:148]: I was worried about this at first, but on second look seems like it's just unused. :+1:

**Review by joestringer — CHANGES_REQUESTED** [review]
Couple of minor comments, otherwise looks good!

File: pkg/policy/repository_test.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/policy/repository_test.go]: So the latter number should be the total number of rules in the repository? And the first number the rules selected for "foo"?

File: pkg/policy/api/egress.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/policy/api/egress.go:184]: Looked into this, and indeed, `canReach*()` functions as well as `GetDestinationEndpointSelectors()` can be removed. I'll prepare a separate commit for this in another PR.

**Review by aanm — APPROVED** [review]

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/policy/repository_test.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/policy/repository_test.go]: Fixed :-)

File: pkg/policy/api/egress.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/policy/api/egress.go:184]: Folded these dead code removals here to avoid unnecessary work fixing the dead functions.

**Comment by jrajahalme** [comment]
Port allocation failure in the Istio Bookinfo test.

**Comment by jrajahalme** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
:shipit:

**Review by ianvernon — COMMENTED** [review]
I'm approving these commits (as proxy and API owners) but I can't approve because this is my PR. But, they are Jarno's commits, so consider this approved by proxy and API codeowners :)

**Pull request merged**